### PR TITLE
Add draft for demonstrating ref-struct-constraint soundness

### PR DESCRIPTION
**Pull request by agocke**
VS Code removed whitespace at end of line, use "hide whitespace" to ignore those changes.

### CODE REVIEW

File: proposals/ref-struct-interfaces.md [hunk 1]
```diff
@@ -116,16 +116,47 @@ T M3<T>(Span<T> span)
...
+  2. Static fields cannot have a ref-struct type.
+  3. Variables of ref-struct type cannot be converted to non-ref structs.
+  4. Variables of ref-struct type cannot be substituted as type arguments.
+  5. Variables of ref-struct type cannot implement interfaces.
+
+The new rules apply to the `heap` constraint:
+
+  1. Fields of classes must have types that satisfy the `heap` constraint.
+  2. Static fields must have types that satisfy the `heap` constraint.
+  3. Variables of types which do not satisfy the `heap` constraint cannot be converted to a type without the `heap` constraint.
```
Comment (CyrusNajmabadi) [inline proposals/ref-struct-interfaces.md:144]: the 'do not', 'cannot' and 'without' are tripping my brain up a lot reading this sentence.  any posibility on simplifying this?

Comment (agocke) [reply proposals/ref-struct-interfaces.md]: I've changed it to something that I believe is equivalent, but I might be missing a case. I would have to double check the spec.

**Review by jaredpar — APPROVED** [review]

**Comment by OJacot-Descombes** [comment]
We already have a `where T : notnull` constraint. Wouldn't a constraint like `where T : allowrefstruct` be more inline with the current constraints without adding a new kind of `allow` constraints or `~` constraints?

**Comment by jaredpar** [comment]
@OJacot-Descombes 

The `where` clause is a _restriction_ on the set of types that can satisfy a generic parameter. This proposal is seeking to add a syntax that expands the set of types that can satisfy a generic parameter. That is why a new syntax is being proposed because it's a significantly different function that what `where` does today.

**Pull request merged**